x-pack/plugins/security_solution/docs/testing/test_plans/detection_response/prebuilt_rules/upgrade_review_algorithms.md

```diff
@@ -0,0 +1,239 @@
+
+**Automation**: 3 integration tests with mock rules + a set of unit tests for each algorithm
+
+```Gherkin
+Given at least 1 installed prebuilt rule is has a new version available
+And the base version of the rule cannot be determined
+And customized <field_name> field is the same as the Elastic update in this upgrade (current version == target version)
+Then for <field_name> field the diff algorithm should output the current version as the merged one without a conflict
+And <field_name> field should not be returned from the `upgrade/_review` API endpoint
+And <field_name> field should not be shown in the upgrade preview UI
+
+CASE: array fields should work the same agnostic of order
+
+Examples:
+| algorithm          | field_name | base_version | current_version | target_version  |
+| single line string | name       | N/A          | "A"             | "A"             |
+| number             | risk_score | N/A          | 1               | 1               |
+| array of scalars   | tags       | N/A          | ["A", "B", "C"] | ["A", "B", "C"] |
+```
+
+### Rule field has an update and a custom value that are NOT the same and the rule base version doesn't exist - `-BC`
+
+#### **Scenario: Rule field is a number or single line string**
+
+**Automation**: 2 integration tests with mock rules + a set of unit tests for the algorithms
+
+```Gherkin
+Given at least 1 installed prebuilt rule is has a new version available
+And the base version of the rule cannot be determined
+And customized <field_name> field is different than the Elastic update in this upgrade (current version != target version)
+Then for <field_name> field the diff algorithm should output the target version as the merged one with a conflict
+And <field_name> field should be returned from the `upgrade/_review` API endpoint
+And <field_name> field should be shown in the upgrade preview UI
+
+Examples:
+| algorithm          | field_name | base_version | current_version | target_version  |
+| single line string | name       | N/A          | "B"             | "C"             |
+| number             | risk_score | N/A          | 2               | 3               |
+```
+
+#### **Scenario: Rule field is an array of scalar values**
+
+**Automation**: 1 integration test with mock rules + a set of unit tests for the algorithm
+
+```Gherkin
+Given <field_name> field is customized by the user (current version != base version)
+And <field_name> field is updated by Elastic in this upgrade (target version != base version)
+And customized <field_name> field is different than the Elastic update in this upgrade (current version != target version)
+Then for <field_name> field the diff algorithm should output a custom merged version with a solveable conflict
+And <field_name> field should be returned from the `upgrade/_review` API endpoint
+And <field_name> field should be shown in the upgrade preview UI
+
+CASE: array fields should work the same agnostic of order
```

Comment:
Does this mean that it's only applicable to some fields and for other fields order doesn't matter?

Reply:
Not sure if I understand the question, things that aren't array fields wouldn't have an order per say. This is basically saying that any fields that are arrays should perform the same no matter the order (e.g. an array of `["A", "B"]` would be treated the same by the algorithm as `["B", "A"]`).

Reply:
Oh, I should have phrased it better. I meant do we have arrays where order matters and arrays where it doesn't?